File: forge/state.py
"""Persistent Forge runtime state."""

from __future__ import annotations

from datetime import UTC, datetime
import json
from pathlib import Path
from typing import Any


def forge_dir(repo: Path) -> Path:
    return repo / ".forge"


def state_path(repo: Path) -> Path:
    return forge_dir(repo) / "state.json"


def ensure_forge_dirs(repo: Path) -> None:
    base = forge_dir(repo)
    for path in [base, base / "memory", base / "audit", base / "worktrees"]:
        path.mkdir(parents=True, exist_ok=True)
    for name, content in _memory_defaults().items():
        target = base / "memory" / name
        if not target.exists():
            target.write_text(content, encoding="utf-8")
    if not state_path(repo).exists():
        write_state(repo, default_state())


def default_state() -> dict[str, Any]:
    return {
        "running": False,
        "last_iteration": "",
        "current_task": None,
        "last_branch": "",
        "last_commit_sha": "",
        "consecutive_failures": 0,
        "cost_today": 0.0,
        "updated_at": now_iso(),
    }


def read_state(repo: Path) -> dict[str, Any]:
    ensure_forge_dirs(repo)
    try:
        data = json.loads(state_path(repo).read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        data = default_state()
    if not isinstance(data, dict):
        return default_state()
    return data


def write_state(repo: Path, data: dict[str, Any]) -> None:
    forge_dir(repo).mkdir(parents=True, exist_ok=True)
    data = dict(data)
    data["updated_at"] = now_iso()
    state_path(repo).write_text(json.dumps(data, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def update_state(repo: Path, **updates: object) -> dict[str, Any]:
    data = read_state(repo)
    data.update(updates)
    write_state(repo, data)
    return data


def stop_file(repo: Path) -> Path:
    return forge_dir(repo) / "STOP"


def request_stop(repo: Path) -> None:
    ensure_forge_dirs(repo)
    stop_file(repo).write_text(f"stop requested at {now_iso()}\n", encoding="utf-8")
    update_state(repo, running=False)


def resume(repo: Path) -> None:
    try:
        stop_file(repo).unlink()
    except FileNotFoundError:
        pass
    update_state(repo, running=False)


def stop_requested(repo: Path) -> bool:
    return stop_file(repo).exists()


def now_iso() -> str:
    return datetime.now(UTC).replace(microsecond=0).isoformat()


def _memory_defaults() -> dict[str, str]:
    return {
        "project_summary.md": "# Project Summary\n\nForge has not summarized this repository yet.\n",
        "architecture_notes.md": "# Architecture Notes\n\n",
        "backlog.json": "[]\n",
        "failed_attempts.json": "[]\n",
        "risky_files.md": "# Risky Files\n\n",
        "test_map.md": "# Test Map\n\n",
        "decisions.md": "# Decisions\n\n",
    }

File: forge-app/src/main/state.mjs
import fs from "node:fs/promises";
import path from "node:path";

export function forgeDir(repo) {
  return path.join(repo, ".forge");
}

export function nowIso() {
  return new Date().toISOString();
}

export async function ensureForgeDirs(repo) {
  const base = forgeDir(repo);
  await fs.mkdir(path.join(base, "audit"), { recursive: true });
  await fs.mkdir(path.join(base, "memory"), { recursive: true });
  await fs.mkdir(path.join(base, "worktrees"), { recursive: true });
  await writeIfMissing(path.join(base, "memory", "backlog.json"), "[]\n");
  await writeIfMissing(path.join(base, "state.json"), `${JSON.stringify(defaultState(), null, 2)}\n`);
}

export async function ensureForgeProjectFiles(repo) {
  await writeIfMissing(path.join(repo, "FORGE_MISSION.md"), "# Forge Mission\n\nLocal autonomous branch factory.\n");
  await writeIfMissing(path.join(repo, "FORGE_BACKLOG.md"), "# Forge Backlog\n\n- [ ] Add the first small safe improvement.\n");
}

export function defaultState() {
  return {
    running: false,
    currentTask: null,
    lastIteration: "",
    lastBranch: "",
    lastCommitSha: "",
    consecutiveFailures: 0,
    costToday: 0,
    updatedAt: nowIso(),
  };
}

export async function readState(repo) {
  await ensureForgeDirs(repo);
  try {
    const parsed = JSON.parse(await fs.readFile(path.join(forgeDir(repo), "state.json"), "utf8"));
    return parsed && typeof parsed === "object" ? parsed : defaultState();
  } catch {
    return defaultState();
  }
}

export async function writeState(repo, updates) {
  const next = { ...(await readState(repo)), ...updates, updatedAt: nowIso() };
  await fs.writeFile(path.join(forgeDir(repo), "state.json"), `${JSON.stringify(next, null, 2)}\n`);
  return next;
}

export async function requestStop(repo) {
  await ensureForgeDirs(repo);
  await fs.writeFile(path.join(forgeDir(repo), "STOP"), `stop requested at ${nowIso()}\n`);
  await writeState(repo, { running: false });
}

export async function resume(repo) {
  await fs.rm(path.join(forgeDir(repo), "STOP"), { force: true });
  await writeState(repo, { running: false });
}

export async function stopRequested(repo) {
  try {
    await fs.access(path.join(forgeDir(repo), "STOP"));
    return true;
  } catch {
    return false;
  }
}

async function writeIfMissing(filePath, content) {
  try {
    await fs.access(filePath);
  } catch {
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, content);
  }
}
